wire-ios-request-strategy/Sources/Request Strategies/Conversation/ConversationParticipantsService.swift

```diff
@@ -0,0 +1,133 @@
+}
+
+class ConversationParticipantsService: ConversationParticipantsServiceInterface {
+
+    // MARK: - Properties
+
+    private let context: NSManagedObjectContext
+    private let proteusParticipantsService: ProteusConversationParticipantsServiceInterface
+    private let mlsParticipantsService: MLSConversationParticipantsServiceInterface
+
+    // MARK: - Life cycle
+
+    convenience init(context: NSManagedObjectContext) {
+        self.init(
+            context: context,
+            proteusParticipantsService: ProteusConversationParticipantsService(context: context),
+            mlsParticipantsService: MLSConversationParticipantsService(context: context)
+        )
+    }
+
+    init(
+        context: NSManagedObjectContext,
+        proteusParticipantsService: ProteusConversationParticipantsServiceInterface,
+        mlsParticipantsService: MLSConversationParticipantsServiceInterface
+    ) {
+        self.context = context
+        self.proteusParticipantsService = proteusParticipantsService
+        self.mlsParticipantsService = mlsParticipantsService
+    }
+
+    // MARK: - Interface
+
+    func addParticipants(
+        _ users: [ZMUser],
+        to conversation: ZMConversation,
+        completion: @escaping AddParticipantAction.ResultHandler
+    ) {
+        guard
+            conversation.conversationType == .group,
+            !users.isEmpty,
+            !users.contains(ZMUser.selfUser(in: context))
+        else {
+            completion(.failure(.invalidOperation))
+            return
+        }
+
+        switch conversation.messageProtocol {
+        case .proteus:
+            proteusParticipantsService.addParticipants(
+                users,
+                to: conversation,
+                completion: completion
+            )
+        case .mls:
+            mlsParticipantsService.addParticipants(
+                users,
+                to: conversation,
+                completion: completion
+            )
+        case .mixed:
+            proteusParticipantsService.addParticipants(
+                users,
+                to: conversation,
+                completion: completion
+            )
+            mlsParticipantsService.addParticipants(
+                users,
+                to: conversation,
+                completion: { _ in }
+            )
+        }
+    }
+
+    func removeParticipant(
+        _ user: ZMUser,
+        from conversation: ZMConversation,
+        completion: @escaping RemoveParticipantAction.ResultHandler
+    ) {
+        guard conversation.conversationType == .group else {
+            return completion(.failure(ConversationRemoveParticipantError.invalidOperation))
+        }
+
+        switch (conversation.messageProtocol, user.isSelfUser) {
+
+        case (.proteus, _), (.mixed, _), (.mls, true):
+            proteusParticipantsService.removeParticipant(
+                user,
+                from: conversation,
+                completion: completion
+            )
```

Comment:
I don't understand why we remove the self user from an mls group via the proteus service.

Reply:
yes me neither, maybe @netbe can comment on that?

Reply:
@netbe bump ⬆️

Reply:
did I write that?

Reply:
actually it was me 😅

Reply:
https://github.com/wireapp/wire-ios-data-model/pull/1409